sub-packages/bionemo-llm/src/bionemo/llm/run/config_models.py

```diff
@@ -0,0 +1,382 @@
+        num_nodes (int): The number of nodes to use. Default is 1.
+
+    Methods:
+        validate_devices(): Validates the number of devices based on the tensor and pipeline model parallel sizes.
+    """
+
+    tensor_model_parallel_size: int = 1
+    pipeline_model_parallel_size: int = 1
+    accumulate_grad_batches: int = 1
+    ddp: Literal["megatron"] = "megatron"
+    remove_unused_parameters: bool = True
+    num_devices: int = 1
+    num_nodes: int = 1
+
+    @model_validator(mode="after")
+    def validate_devices(self):
+        """Validates the number of devices based on the tensor and pipeline model parallel sizes."""
+        if self.num_devices < self.tensor_model_parallel_size * self.pipeline_model_parallel_size:
+            raise ValidationError(
+                "devices must be divisible by tensor_model_parallel_size * pipeline_model_parallel_size"
+            )
+        return self
+
+
+class TrainingConfig(BaseModel):
+    """TrainingConfig is a configuration class for training models.
+
+    Attributes:
+        max_steps (int): The maximum number of training steps.
+        limit_val_batches (int | float): The number of validation batches to use. Can be a fraction or a count.
+        val_check_interval (int): The interval (in steps) at which to check validation.
+        precision (Literal["32", "bf16-mixed", "16-mixed"], optional): The precision to use for training. Defaults to "bf16-mixed".
+        accelerator (str, optional): The type of accelerator to use for training. Defaults to "gpu".
+        gc_interval (int, optional): The interval of global steps at which to run synchronized garbage collection. Useful for synchronizing garbage collection when performing distributed training. Defaults to 0.
+    """
+
+    max_steps: int
+    limit_val_batches: int | float  # Because this can be a fraction or a count...
+    val_check_interval: int
+    precision: Literal["32", "bf16-mixed", "16-mixed"] = "bf16-mixed"
+    accelerator: str = "gpu"
+    # NOTE: VERY important for distributed training performance.
+    gc_interval: int = 0
+
+
+class OptimizerSchedulerConfig(BaseModel):
+    """Configuration for the optimizer and learning rate scheduler.
+
+    Attributes:
+        lr (float): Learning rate for the optimizer. Default is 1e-4.
+        optimizer (str): Type of optimizer to use. Default is "adam".
+        cosine_rampup_frac (float): Fraction of total training steps for the cosine ramp-up phase. Default is 0.01.
+        cosine_hold_frac (float): Fraction of total training steps to hold the learning rate constant after ramp-up. Default is 0.05.
+        interval (str): Interval for updating the learning rate scheduler. Default is "step".
+        monitor (str): Metric to monitor for learning rate adjustments. Default is "val_loss".
+    """
+
+    lr: float = 1e-4
+    optimizer: str = "adam"
+    cosine_rampup_frac: float = 0.01
+    cosine_hold_frac: float = 0.05
```

Comment:
Maybe not cosine annealing since this is specific to geneformer.

Reply:
gave it a stop gap fix, need to follopw up with a real fix asap. This is a pretty hard problem since the set of parameters + schedulers is unbounded.